### ISSUE TITLE
Duplicated OperationID when adding route with multiple methods

### ISSUE THREAD

**Issue by bruchar1** [tiangolo/fastapi#4740]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
router.add_api_route(
    "/clear",
    clear,
    methods=["POST", "DELETE"]
)
```

### Description

Since to be caused by #4650.

The new `generate_unique_id()` function uses `list(route.methods)[0].lower()` as suffix for the `operation_id`. Therefore, in my example, both post and delete endpoints get `_post` suffix for operation_id, causing it to no longer be unique.

It then issues a "UserWarning: Duplicate Operation ID"

### Operating System

Windows

### FastAPI Version

0.75.0

### Python Version

3.10.2

**Comment by mbrav** [comment]
I am getting a similar problem. Am getting the following errors.

```log
File "/fastapi/applications.py", line 216, in openapi
    return JSONResponse(self.openapi())
  File "/fastapi/applications.py", line 191, in openapi
    self.openapi_schema = get_openapi(
  File "/fastapi/openapi/utils.py", line 414, in get_openapi
    definitions = get_model_definitions(
  File "/fastapi/utils.py", line 35, in get_model_definitions
    model_name = model_name_map[model]
```

Upon further inspection of [the code](https://github.com/tiangolo/fastapi/blob/26f725d259c5dbe3654f221e608b14412c6b40da/fastapi/openapi/utils.py#L412-L416=) with the following lines:


```python
flat_models = get_flat_models_from_routes(routes)
model_name_map = get_model_name_map(flat_models)
definitions = get_model_definitions(
    flat_models=flat_models, model_name_map=model_name_map
)
```

The problem is that basically that in my case, the variables passed into `get_model_definitions()` are **flat_models** where `len(flat_models)=16` and **model_name_map** where `len(flat_models)=15`, hence the error. Here is a screen from debugger:

![Screenshot_20220402_225519](https://user-images.githubusercontent.com/7410291/161399211-622767a1-3088-405f-b288-77ea83ebbcde.png)

So the problem in my case is that `<class 'pydantic.main.Body_access_token_login_api_auth_token_post'>` gets duplicated twice in **flat_models** and both for some reason seem to have the same hashes. For more info, it seems to be the **fastapi.security.OAuth2PasswordRequestForm** that is causing the trouble, but no sure. 

Once I remove code with my auth endpoints, openapi schema gets generated and docs do not crash, but I get the following error. 

```log
/fastapi/openapi/utils.py:183: UserWarning: Duplicate Operation ID tasks_list_api_tasks_get for function tasks_list at ./app/api/tasks.py
  warnings.warn(message)
```

My repo with the following problem is referenced above.

**Comment by mbrav** [comment]
Nevermind, found a solution. For future reference, the commit above  fixed the issue. I was transferring some code from another project. Sorry if maybe not exactly relevant to the issue, but hope it helps someone in the future.

**Comment by jonatasoli** [comment]
@bruchar1 this solution above solve yout problem?

**Comment by bruchar1** [comment]
No. @mbrav problem is unrelated to this.

**Comment by jonatasoli** [comment]
> Nevermind, found a solution. For future reference, the commit above fixed the issue. I was transferring some code from another project. Sorry if maybe not exactly relevant to the issue, but hope it helps someone in the future.

@mbrav I believe this issue can be closed then :)

**Comment by mbrav** [comment]
@jonatasoli I agree, but I am not the one who opened it :)

**Comment by bruchar1** [comment]
@jonatasoli Why do you want to close this? The bug is still there, and completely unrelated to @mbrav issue.

**Comment by bneijt** [comment]
There are multiple issues here:
1) The [APIRoute class](https://github.com/tiangolo/fastapi/blob/master/fastapi/routing.py#L307) generates the unique id but also supports multiple methods. This should only support a single method if it wants to use the method to generate a unique id (which to me seems like the only way to generate them, maybe a random uuid with proper seeding might also work, but that is probably hard to work with)
2) The methods accessed by the [generate_unique_id](https://github.com/tiangolo/fastapi/blob/master/fastapi/utils.py#L141) function are modeled as a `set`, which makes the order unpredictable. So the `list(route.methods)[0]` will result in different values for different runs, making the operationid unstable.

I'll create a quick MR to change the `list` to a `sorted`, but I think we should also change the APIRoute class to only accept a single method. Before going down that refactoring, I would like to see a response from any of the maintainers.

**Comment by stapetro** [comment]
Hello @bneijt, I confirm I hit the same issue with fastapi@0.79.1. I've decorated a method like this:
```python
@router.api_route(
    "/example",
    methods=["POST", "PUT"],
    status_code=status.HTTP_201_CREATED,
)
async def save_example(...)
  pass
```

**Comment by 462548187** [comment]
health_system.py
===================================================================================
router = APIRouter(prefix="/system", route_class=ContextLogerRoute)


@router.get('/health', tags=['默认开启的健康检查'])
async def health_check():
    """
    注册全局健康检查
    :return:
    """
    return Success(data='ok')

===============================================================================================
wanring: /Users/.../..../venv/lib/python3.11/site-packages/fastapi/openapi/utils.py:184: UserWarning: Duplicate Operation ID health_check_api_system_health_get for function health_check at /Users/.../..../fastApiFramework/app/routers/health/health_system.py
  warnings.warn(message)
=====================================

How to solve this problem?

**Comment by fulanii** [comment]
### Similar problem : `UserWarning: Duplicate Operation ID` for all my routes

I'm getting this problem with all my routes, looked online and in documentation cant find a fix for it. Can anyone help solve this?

I have these methods and endpoints in my `users.py`
<img width="832" alt="Screenshot 2022-12-17 at 5 52 35 PM" src="https://user-images.githubusercontent.com/68094236/208270488-d27b8d60-5a8e-45da-a854-c8804ee34d5c.png">

`posts.py`
<img width="928" alt="Screenshot 2022-12-17 at 5 53 02 PM" src="https://user-images.githubusercontent.com/68094236/208270517-691d5697-1398-4fa1-8e6a-8a19a5cccf59.png">

and im getting this in the terminal
<img width="1634" alt="Screenshot 2022-12-17 at 5 55 35 PM" src="https://user-images.githubusercontent.com/68094236/208270554-60b7d4a0-1c5f-42ff-8673-d5fe5031a791.png">

**Comment by ricardo-reis-1970** [comment]
# Similar issue: UserWarning: Duplicate Operation ID

I've been reducing my example until all I have is this:
## Freeze
fastapi==0.88.0
uvicorn==0.20.0
*NOT* using libs from global environment.

## `users.py`
```python
from fastapi import APIRouter

router = APIRouter()

@router.get("")
def users():
    return 1
```
## `__init__.py`
```python
# typical boilerplate

@app.get("/get_config")
def get_config():
    return cfg.properties
```
## `main.py`
```python
from <my_package> import app
...
app.include_router(users.router, prefix="/users", tags=["Users"])
```

With this, I get the warning:
```bash
[...].venv/lib/python3.10/site-packages/fastapi/openapi/utils.py:184: UserWarning: Duplicate Operation ID users_users_get for function users at [...]/<my_library>/routers/users.py
  warnings.warn(message)
```

# Funny, though...
... when I take the endpoint from an `app.route` mount and paste it directly into `__init__.py`, like this:
```python
app.get("/users")
def users():
    return 1
```
, just adjusting the endpoint, so that it is the same `/users`, no warning is issued!

**What am I missing? Am I using `app.include_router` wrong?** The API seems to work anyway, but some of us just don't like warnings, because those make us think we're not perfect.

**Comment by bkis** [comment]
@ricardo-reis-1970 do you get these warnings only when running tests or also when running the app itself? Because for me it's only when running pytest. I haven't session-scoped my `test_app` or `test_client` fixtures so it's creating new instances with every test function (on purpose). Maybe that hints to the underlying issue?

**Comment by KRC1997** [comment]
Just a thought: Can you or anyone facing the issue, confirm if you have included the router (with all routes) only once in the parent router or the fastapi app?
I was facing the same warning when I noticed I had added called `app.include_router` with the same router argument more than once. It resolved itself when I removed the extra call.

**Comment by bkis** [comment]
@KRC1997  Yeah you're right (in a way). I only add the router once, but the problem occurred when I started using [asgi-lifespan](https://pypi.org/project/asgi-lifespan/0.5.0/) to include the startup- and shutdown routines (the app lifecycle) of my app when running tests. I had my `test_app` fixture not session-scoped but test-scoped (default) so it was re-initializing the app (including startup routines) for each and every test. But of course the actual `app: FastAPI` object was only imported once, so the startup routine was applied to the same app instance for every test. Now when you add the routers in that very startup routine, it will be added every time.

The following workaround fixed the problem for me and also fixed another issue I had anyway: I now still initialize the database client in the startup routine (and close it in the shutdown routine) but everything that belongs to initializing the actual app instance (e.g. registering routers), I moved to a `pre_startup` routine/function that just runs whenever the app module is loaded (on import):

```python
def pre_startup_routine(app: FastAPI) -> None:
    # add middlewares
    ...

    # register routers
    ...

async def startup_routine() -> None:
    # init database client
    ...

async def shutdown_routine() -> None:
    # close database client and cleanup
    ...

# create FastAPI app instance
app = FastAPI(
    # ...
    on_startup=[startup_routine],
    on_shutdown=[shutdown_routine],
)

pre_startup_routine(app)
``` 

This has the additional advantage that it works well with my test setup.
If the reason for this warning is really only this mistake on the dev side, I guess this issue might also be closed as it's not really an issue with FastAPI but rather with how the app's lifecycle is handled in some situations.
